src/frontend/src/expr/type_inference/func.rs

```diff
@@ -19,13 +19,27 @@ use num_integer::Integer as _;
 use risingwave_common::error::{ErrorCode, Result};
 use risingwave_common::types::{DataType, DataTypeName};
 
-use super::{align_types, cast_ok_base, CastContext};
+use super::{align_types, cast_ok_base, CastContext, least_restrictive};
 use crate::expr::{Expr as _, ExprImpl, ExprType};
 
 /// Infers the return type of a function. Returns `Err` if the function with specified data types
 /// is not supported on backend.
 pub fn infer_type(func_type: ExprType, inputs: &mut Vec<ExprImpl>) -> Result<DataType> {
     if let Some(res) = infer_type_for_special(func_type, inputs).transpose() {
+        // cast to res here? 
+        let res_type = res.clone().unwrap();
+
+        let inputs_owned = std::mem::take(inputs);
+        *inputs = inputs_owned
```

Comment:
I have the correct inferred return type (`numeric` in the cases below). How do I cast the inputs to this type? With the code here I get

```SQL
select 1::int || array[1.2]::numeric[];
ERROR:  QueryError: Bind error: cannot cast type "numeric[]" to "numeric" in Implicit context

select array[1.2]::numeric[] || 1::int;
ERROR:  QueryError: Bind error: cannot cast type "numeric[]" to "numeric" in Implicit context
```

I tried to avoid above error by only casting numeric values. I then receive